File: solutions/python/rational-numbers/1/rational_numbers.py
from __future__ import division


class Rational(object):
    def __init__(self, numerator, denominator):
        if denominator == 0:
            raise ValueError("Denominator cannot be 0")
        if not (isinstance(numerator, int) and isinstance(denominator, int)):
            raise ValueError("Numerator and Denominator have to be integers")
        self.numerator = numerator
        self.denominator = denominator
        self.reduce()

    @staticmethod
    def gcd(a, b):
        if a == 0:
            return b
        if b == 0:
            return a
        return Rational.gcd(b, a % b)

    def reduce(self):
        div = Rational.gcd(self.numerator, self.denominator)
        self.numerator //= div
        self.denominator //= div
        return self

    def __eq__(self, other):
        return (
            self.numerator == other.numerator
            and self.denominator == other.denominator
        )

    def __repr__(self):
        return f"{self.numerator}/{self.denominator}"

    def __add__(self, other):
        return Rational(
            self.numerator * other.denominator
            + other.numerator * self.denominator,
            (self.denominator * other.denominator),
        )

    def __sub__(self, other):
        return Rational(
            self.numerator * other.denominator
            - other.numerator * self.denominator,
            (self.denominator * other.denominator),
        )

    def __mul__(self, other):
        return Rational(
            self.numerator * other.numerator,
            self.denominator * other.denominator,
        )

    def __truediv__(self, other):
        return Rational(
            self.numerator * other.denominator,
            self.denominator * other.numerator,
        )

    def __abs__(self):
        if self.numerator < 0:
            self.numerator *= -1
        if self.denominator < 0:
            self.denominator *= -1
        return self

    def __pow__(self, power):
        if power > 0:
            return Rational(self.numerator**power, self.denominator**power)
        return Rational(self.denominator**-power, self.numerator**-power)

    def __rpow__(self, base):
        return pow(base**self.numerator, 1 / self.denominator)

File: solutions/javascript/rational-numbers/1/rational-numbers.js
/**
 * Class representing a rational number.
 */
class Rational {
  /**
   * Create a rational number.
   * @param {number} numerator - The numerator of the rational number.
   * @param {number} denominator - The denominator of the rational number.
   * @throws Will throw an error if the denominator is zero.
   */
  constructor(numerator, denominator) {
    if (denominator === 0) {
      throw new Error('Denominator must not be zero.');
    }

    this.numerator = numerator;
    this.denominator = denominator;

    this.reduce();
    this.ensureSignInNumerator();
  }

  /**
   * Add two rational numbers.
   * @param {Rational} that - The rational number to add.
   * @returns {Rational} The sum of the two rational numbers.
   */
  add(that) {
    const commonDenominator = this.denominator * that.denominator;
    return new Rational(
      this.numerator * that.denominator + that.numerator * this.denominator,
      commonDenominator,
    );
  }

  /**
   * Subtract one rational number from another.
   * @param {Rational} that - The rational number to subtract.
   * @returns {Rational} The difference of the two rational numbers.
   */
  sub(that) {
    const commonDenominator = this.denominator * that.denominator;
    return new Rational(
      this.numerator * that.denominator - that.numerator * this.denominator,
      commonDenominator,
    );
  }

  /**
   * Multiply two rational numbers.
   * @param {Rational} that - The rational number to multiply.
   * @returns {Rational} The product of the two rational numbers.
   */
  mul(that) {
    return new Rational(
      this.numerator * that.numerator,
      this.denominator * that.denominator,
    );
  }

  /**
   * Divide one rational number by another.
   * @param {Rational} that - The rational number to divide by.
   * @returns {Rational} The quotient of the two rational numbers.
   */
  div(that) {
    return new Rational(
      this.numerator * that.denominator,
      this.denominator * that.numerator,
    );
  }

  /**
   * Get the absolute value of the rational number.
   * @returns {Rational} The absolute value of the rational number.
   */
  abs() {
    return new Rational(Math.abs(this.numerator), Math.abs(this.denominator));
  }

  /**
   * Raise the rational number to an integer power.
   * @param {number} n - The exponent.
   * @returns {Rational} The rational number raised to the power n.
   */
  exprational(n) {
    return new Rational(
      Math.pow(this.numerator, n),
      Math.pow(this.denominator, n),
    );
  }

  /**
   * Raise a real number to the power of the rational number.
   * @param {number} base - The base real number.
   * @returns {number} The result of the base raised to the power of the rational number.
   */
  expreal(base) {
    return Math.pow(
      10.0,
      Math.log10(Math.pow(base, this.numerator)) / this.denominator,
    );
  }

  /**
   * Reduce the rational number to its simplest form.
   * @returns {Rational} The reduced rational number.
   */
  reduce() {
    const commonDivisor = this.gcd(this.numerator, this.denominator);

    this.numerator /= commonDivisor;
    this.denominator /= commonDivisor;
    this.ensureSignInNumerator();

    return this;
  }

  /**
   * Calculate the greatest common divisor of two numbers.
   * @param {number} a - The first number.
   * @param {number} b - The second number.
   * @returns {number} The greatest common divisor of a and b.
   */
  gcd(a, b) {
    let localA = a;
    let localB = b;
    while (localB !== 0) {
      const t = localB;
      localB = localA % localB;
      localA = t;
    }
    return localA;
  }

  /**
   * Ensure the sign is in the numerator.
   */
  ensureSignInNumerator() {
    if (this.denominator < 0) {
      this.denominator = -this.denominator;
      this.numerator = -this.numerator;
    }
  }
}

export { Rational };
